### DISCUSSION TITLE
How should logging work?

### DISCUSSION THREAD

**Discussion by gbishop** [UNC-Project-Open-AAC/OS-DPI#168] [Ideas]
As currently implemented logging requires adding a Logging component to the design and then updating its state variable in Actions. This has several issues:

1. The syntax for the update is cumbersome and error prone. For example if you want to log several state variables you have to write Log("label1", $var1, "label2", $var2, "label3", $var3).
2. You have to do this in every action in which you want the values logged.
3. The values you are logging are those **before** the action completes. So if you log the display you are saving its old value, not the new value you might expect. 
4. After you have configured this there is no way to turn logging off and on. I worry that the logs are going to be forgotten and grow until they fill the database.

**Proposal**

I suggest we put this complexity into the Logger component. 

* A list of the $States and #Fields to log.
* A checkbox to enable logging (I suggest this defaults to off when importing a design).
* Some way to limit the log size:
  * A number specifying the maximum number of log records to save, or
  * a date to stop logging, or
  * some combination of time (keep logs for the most recent N days) and counts (keep the last N logs).
* A checkbox to make logging selectable for individual actions. This could add a checkbox to each action so you could log some but not others. 

We could still support adding a string to the log by assigning to the $log state variable but I'm guessing this would be less frequently used. We could also allow enabling and disabling logging from Actions so you could have a button in the UI to turn it on and off. 

What do you all think?

**Comment by kaericks** [comment]
Jumping in to keep the thread going... Putting the complexity into the Logger component sounds like a good solution. Lists and checkboxes will certainly make all of the options easier for people like me to implement.  As for limiting the log size, forcing the selection of time (N days or a date to stop logging - with a forced limit) sounds like a good solution. Is it possible to consider a way to auto-download/export logs?  In the Qualtrics environment where we do lots of our professional development work, we can set the option for it to send us the data every N days. Maybe an option to download logs on demand and/or get logs automatically downloaded and sent before they are dumped?

**Comment by gbishop** [comment]
Thanks.

Automatically downloading or exporting would be great but requires
interacting with a system file-save dialog because of security. This is
something that I imagine would be difficult for a user but maybe users of
these experimental interfaces always have someone there to help?

We could overcome this and other browser limitations by embedding our app
in a native executable. I've never done it but it is claimed to be
straightforward.

We could automatically post the log to a web server but that brings up
privacy concerns. We've been avoiding those by keeping everything on the
local computer. As soon as we start posting private data we're into a
different world of security.

gb

**Reply by cdsjeff** [reply]
I like the proposal so far. I assume that each item will also contain a time stamp, preferably with a resolution of .1 sec, or user settable. Also the running clock time is fine for me, but may pose a problem for someone not use to seeing a long string of numbers.

I’m fine with providing the ability to limit the time stamps. But I’ve done studies that last over a week and having the ability to keep logging running for a long period of time can be vital to a study. Maybe provide that as an alternative with a warning.

I will look at what we did 25 years ago when we developed the universal logging format and think about anything additional we need.  now that screens are dynamic, It may be valuable to think about recording cursor position, and cursor position in the message display window.  For example if we are trying to build an osdpi interface to facilitate inline editing (I.e., self-repair) during conversation, it would be important to know where the cursor was placed and the surrounding context.

When we were doing our log file work a bunch of years ago, we were able to feed the log file into a revised version of our interface and essentially “play back” a version of the device actions on the interface.  I’m not asking for this but just hallucinating the possibility of doing something like this in the future.  

Jeff

**Reply by gbishop** [reply]
Yes, date and time on every entry.

Yes, unlimited will be one of the choices. My guess is the maximum size supported by the browser is REALLY large.

Logs could (I think) include pointer coordinates and any cursor position. Summary statistics such as number of button hits or distance the pointer moved could be included. Recording every mouse movement could get length but could be done.

Playing back these log files should be easy if you record all of the state. In our system, the state determines everything on the screen. Visual cues (the current button in a two-switch scan for example) are not currently included in the state but they could be. 

**That reminds me of @sofia-bg  question about nested conversations.** For example, you're composing a response and someone new comes in so you want to briefly put your response aside so you can address them. It should be possible to save the current state and restore it later. This could be a stack for multiple nesting though people often get confused by stacks. A save/restore button on a UI should be doable.

**Reply by cdsjeff** [reply]
Sorry, I profess extreme ignorance here, but: can osdpi distinguish between button types (e.g., letter, word, navigation)?  If so, could that be logged as well as the specific content of the button?  Also it would  be important to be able to attach the current message state, as well as the contents of the message display when it is selected / spoken.
J

**Reply by sofia-bg** [reply]
@cdsjeff presumably it could distinguish between ways you marked the content related to the button in your spreadsheet. So maybe you add a column to the spreadsheet that is "button_type" and then label the various content that way based on the type of content in the cell? Then tell the logger to give you the $button_type for all items logged?

**Reply by cdsjeff** [reply]
That sounds good.  The summary stats are ok, but it may be better to do that as an analysis component.  We could contribute that with a companion analysis program.  That’s on our list of things to do, and I would want to build one with broad analysis features, not just the ones we want for today.  We developed a really nice one a number of years ago  - AQUA - of which I can still get running (windows 97 or windows 7).  I can explore the features with that.   We should confer with Lori and Karen about the approach and feature set.
J

**Reply by cdsjeff** [reply]
Sophia,
That is a great idea. Better than hard-coding it in.
J

**Reply by cdsjeff** [reply]
I’ll improve my posting educate.

**Comment by gbishop** [comment]
Sofia is correct. Put whatever you want in the content spreadsheet and it
will be logged.

**Comment by gbishop** [comment]
An analysis component is the right way to approach it

**Reply by cdsjeff** [reply]
@sofia-bg @kaericks : it would not be too hard to integrate the data log into elan, which could make our transcriptions faster and better if you are interested in calturing device actions.

**Reply by sofia-bg** [reply]
that could certainly be interesting to explore

**Comment by su-aac** [comment]
[](url)
[AAC Data Logging Notes.docx](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/10883949/AAC.Data.Logging.Notes.docx)

I uploaded a word document about what I have serached on data logging, including what data has been logged, what was analyzed,  the challenges of language sample collection in people who use AAC,and privacy concerns. I also summarized here.

What data has been logged:  
- Timestamp 
- Content output 
- Access: e.g. touchscreen, mouse, joystick, eye tracker, etc. 
- Source: type of source that produces an entry. e.g. word list, pages, commands, etc. 
- Action on keyboard: e.g. enter, select, delete 
- Coordination: coordinate position of an entry 
- Position label: label of the position 
- Page label: name of the page 

What was analyzed?
- Total number of words
- Number of unique words
- Type token ratio
- Word count
- Word type count
- Communication rate
- Selection rate

Challenges of  language sample collection and analysis in people who use AAC (Mooney et al., 2020)
- Vocabulary access is limited by the vocabulary available on their communication devices 
- Length of time is long.
- Analysis Procedures lack of standard 
- Differences across access methods and how to consider the use of word predictive features 
- Filtering modeling from a third-party helper (Cross & Segalman, 2016)- Andrea mentioned the feature of an additional hit “MODEL“ before modeling

Privacy Concern ([AssistiveWare](https://www.assistiveware.com/blog/aac-data-collection-and-privacy#:~:text=AssistiveWare%20does%20not%20include%20LAM,without%20invading%20the%20user's%20privacy))
- AAC data collection principles
•	Only collect data if there is a clear purpose. 
•	Ask for permission to collect data.  
•	Only collect anonymous data.  
•	Do not collect more details than necessary.
•	Do not collect sensitive personal data.
•	Do not share raw data with third parties.  
- Solutions they proposed:  
•	Rather than collecting the real content of words, we can collect…types of words: e.g. total number of nouns, verbs, and adjectives; compare total number of words for weekdays with weekend days (to compare school and home use) 
•	Sophie’ idea: recording signs show up on the screen (like the red dot during Zoom recording)

**Reply by lageist** [reply]
Thanks for this summary @hsiaotingsu

**Reply by aetkie** [reply]
@hsiaotingsu thanks for these notes! Very helpful. I'm wondering if you could differentiate between Communication rate and 
Selection rate?
I would be interested in thinking of a way to log selection rate (maybe with a start/stop time frame, like, user is on the page and the clock starts and then stops when user selects pre-determined symbol). But what would you log/measure with communication rate? Is this words/symbols/utterances across a certain amount of time (wpm?)

**Reply by su-aac** [reply]
@aetkie Hi Andrea!

Below are the definitions of the communication rate and selection rate from Chen, Wadhwa, and Nyberg (2019).
- Average Communication Rate: number of words generated by the user per minute.
- Average Selection Rate: number of selection actions taken per minute by the user while communicating via the device. 

The way you proposed (the user is on the page, and the clock starts and then stops when the user selects a pre-determined symbol) is valuable and should be measured too!

**Reply by aetkie** [reply]
Thanks for sharing those definitions @hsiaotingsu

**Reply by cdsjeff** [reply]
Thanks @hsiaotingsu!  I think the base measures of communication rate / words per minute are pretty straightforward (with some variations), but the implementation is more complex:
- If a person is in a conversation, what portion of the logfile do you use for analysis? If you use the whole thing, then you are including the other persons talking time, down time, etc.
- If you use it in the field (e.g., participant uses device for a week) how do you decide when a communication begins, how much inactivity constitutes a pause or stop, what is an utterance versus just playing around?

I think these issues, which are real for any research project, go beyond the logfile and should be dealt with using an analysis program of some sort. I think that for any measure of communication rate, one will need to select the initiating event (e.g., first relevant selection) and the end event (e.g., speech output).  Otherwise you will need to implement some automated procedure that filters out errant selections, etc.  We did some work on that about 20 years ago. I'm attaching an article we got published on that.
[Smith et al. 2006 - The development of an automated method for analyzing communication rate in augmentative and alternative communication.pdf](https://github.com/UNC-Project-Open-AAC/OS-DPI/files/11055772/Smith.et.al.2006.-.The.development.of.an.automated.method.for.analyzing.communication.rate.in.augmentative.and.alternative.communication.pdf)

**Reply by aetkie** [reply]
@cdsjeff thanks for linking that article. There are definitely more complex applications of logging when thinking of a single person with limited targets and compared to a person in a conversation. I am thinking about designing a UI for my dissertation research and how I would use logging to measure latency and accuracy of the participants. Much more simple than conversational communication rate but needed some definitions to get me thinking of methods.

**Comment by gbishop** [comment]
I've got a first version of logging ready to push to production. You have a state variable ($Log is the default) to let you turn logging on and off. There is a date field which will turn off logging after the given date; set this far into the future to disable. Then you provide a list of state variables and fields you want logged. It just wants a space separated list of names and can continue onto 
multiple lines.

There is a small slowly pulsing yellow indicator in the upper right of the screen when logging is active. You can change its position and appearance using the customize component. 

![Screenshot from 2023-03-06 13-55-52](https://user-images.githubusercontent.com/198012/223204621-71104818-633b-4117-9289-0379e1bd342e.png)

**Comment by cdsjeff** [comment]
One area that I think would be productive to measure is how much someone participates on their device by hour of day and day of the week. Here is a dot-plot graph representing two weeks of logfiling. each little x represents the number of words spoken in an hour.  You can see that this individual (in high school using a dynavox) produced more talk in the morning and later afternoon than during most of his school day.  Varies by individual a bit.  I think our logfile program could be important for providing data for this kind of analysis.
![](https://previews.dropbox.com/p/thumb/AB2F7BDAqXfJ07MLjPGUezBd8o9BhRZ3aqGaObyl_vG9dUKqdsFaQyNFneJ9ksKxZFba5W_cliUWDOYJMdHfgoQ0BD81V-g2J3EATZxP7EuKQvzPDoOQEUttNhZ40It9WIJiTeuQxPyRFmlrJtl5FqAfw5bpmNmpL_yci60VwNrv-BVnUa-LAuaJIgJhiTE3s-lzv_QeIZ99rdd11lC1_Gvz2ph0qwzRlyPu8I-iaUiwR_KLt6YxiQrIL02E4qAzS7xzNc6UwUabJfVvBpUvdedFUMTVsBlhtp-XGgWvmZXz-cLFXCBaZ3Mxztq4Q7d8ALHichKevsf7hxIqZfst-cLYG4dvobYZZsLjIVXirlV1Z0PGa7ljvgDJwH1qoVpPUlc/p.png)

**Reply by cdsjeff** [reply]
The graph didn't upload. Here's my second attempt.
<img width="645" alt="Screen Shot 2023-03-23 at 4 40 07 PM" src="https://user-images.githubusercontent.com/32512117/227353514-bf1ef3fe-f35a-4c86-80e6-6bb6a47184c7.png">

**Reply by cdsjeff** [reply]
This represents hours per day for 2 5-day (m-f) week periods collapsed across days.

**Reply by aetkie** [reply]
OS-DPI could definitely log this data
